### ISSUE TITLE
Example doesnt load model and missing sky.png, ground.png and water.png

### ISSUE THREAD

**Issue by Meai** [flowtsohg/mdx-m3-viewer#14]
Hi, I'm trying to get the examples/index.html to display any kind of .mdx model.
It doesnt even show the supplied. obj model.
I added .mdx as supported filetype directly, I added some dummy .pngs for the stuff I mentioned in the title but all it shows is the ground plane. No model.
I did the compilation via ruby, but still.. I assume this is just outdated?
Any clue what's going wrong, because after I added .mdx to the supportedFileTypes array I'm not getting any more errors.

I'm trying this with AzureDragon.mdx

**Comment by flowtsohg** [comment]
I kind of left the example handlers outdated for a while, so I am not sure if they work.
You have to add mdx/m3 manually, this is because they are no longer required for the viewer (the code base went over major rewrites a couple of times, so a lot of stuff changed).
However, by manually I mean you should call Viewer.registerModelHandler.
E.g.:
`viewer.registerModelHandler(".mdx", Mdx.Model, Mdx.ModelInstance, 1);`

Either way, it works fine for me, can you show your client code?

**Comment by Meai** [comment]
Thanks for the response, I closed the issue because I did get it working after lots of debugging. Btw, for my local codebase I removed the ruby compilation step and now just work with whole files, I think that's way easier to understand. I also use typescript for the shaders like this: (es6 template strings)
Both Visual Studio and Webstorm now compile that to normal js pretty much on the fly.

```javascript
var SHADERS:any = {}
SHADERS.decodefloat = `
vec2 decodeFloat2(float f) {
    vec2 v;

    v[1] = floor(f / 256.0);
    v[0] = floor(f - v[1] * 256.0);

    return v;
}

vec3 decodeFloat3(float f) {
    vec3 v;

    v[2] = floor(f / 65536.0);
    v[1] = floor((f - v[2] * 65536.0) / 256.0);
    v[0] = floor(f - v[2] * 65536.0 - v[1] * 256.0);

    return v;
}
`;

SHADERS.pscolor = `
uniform vec3 u_color;

void main() {
    gl_FragColor = vec4(u_color / 255.0, 1);
}
`;

SHADERS.pstexture = `
uniform sampler2D u_texture;

varying vec2 v_uv;

void main() {
    gl_FragColor = texture2D(u_texture, v_uv);
}
`;

SHADERS.pswhite = `
void main() {
    gl_FragColor = vec4(1);
}
`;

SHADERS.psworld = `
uniform sampler2D u_texture;
uniform float u_a;

varying vec2 v_uv;

void main() {
    gl_FragColor = vec4(texture2D(u_texture, v_uv).rgb, u_a);
}
`;

SHADERS.vsbonetexture = `
uniform sampler2D u_boneMap;
uniform float u_matrix_size;
uniform float u_texel_size;

mat4 boneAtIndex(float index) {
    float offset = index * u_matrix_size;
    return mat4(texture2D(u_boneMap, vec2(offset, 0)), texture2D(u_boneMap, vec2(offset + u_texel_size, 0)), texture2D(u_boneMap, vec2(offset + u_texel_size * 2.0, 0)), texture2D(u_boneMap, vec2(offset + u_texel_size * 3.0, 0)));
}
`;

SHADERS.vstexture = `
uniform mat4 u_mvp;

attribute vec3 a_position;
attribute vec2 a_uv;

varying vec2 v_uv;

void main() {
    v_uv = a_uv;

    gl_Position = u_mvp * vec4(a_position, 1);
}
`;

SHADERS.vswhite = `
uniform mat4 u_mvp;

attribute vec3 a_position;

void main() {
    gl_Position = u_mvp * vec4(a_position, 1);
}
`;

SHADERS.vsworld = `
uniform mat4 u_mvp;
uniform vec2 u_uv_offset;

attribute vec3 a_position;
attribute vec2 a_uv;

varying vec2 v_uv;

void main() {
    v_uv = a_uv + u_uv_offset;

    gl_Position = u_mvp * vec4(a_position, 1);
}
`


// MDX

SHADERS.wpsmain = `
uniform sampler2D u_texture;
uniform bool u_alphaTest;
uniform vec4 u_modifier;

varying vec3 v_normal;
varying vec2 v_uv;

void main() {
    #ifdef STANDARD_PASS
    vec4 texel = texture2D(u_texture, v_uv).bgra;

    // 1bit Alpha
    if (u_alphaTest && texel.a < 0.75) {
        discard;
    }

    gl_FragColor = texel * u_modifier;
    #endif

    #ifdef UVS_PASS
    gl_FragColor = vec4(v_uv, 0.0, 1.0);
    #endif

    #ifdef NORMALS_PASS
    gl_FragColor = vec4(v_normal, 1.0);
    #endif

    #ifdef WHITE_PASS
    gl_FragColor = vec4(1.0);
    #endif
}
`

SHADERS.wpsparticles = `
uniform sampler2D u_texture;

varying vec2 v_uv;
varying vec4 v_color;

void main() {
    gl_FragColor = texture2D(u_texture, v_uv).bgra * v_color;
}
`
SHADERS.wvscolor = `
uniform mat4 u_mvp;

attribute vec3 a_position;
attribute vec4 a_bones;
attribute float a_bone_number;

void main() {
    vec4 v = vec4(a_position, 1);
    vec4 p = (boneAtIndex(a_bones[0]) * v + boneAtIndex(a_bones[1]) * v + boneAtIndex(a_bones[2]) * v + boneAtIndex(a_bones[3]) * v) / a_bone_number;

    gl_Position = u_mvp * p ;
}
`

SHADERS.wvsmain = `
uniform mat4 u_mvp;
uniform vec3 u_uv_offset;

attribute vec3 a_position;
attribute vec3 a_normal;
attribute vec2 a_uv;
attribute vec4 a_bones;
attribute float a_bone_number;

varying vec3 v_normal;
varying vec2 v_uv;

void transform(vec3 inposition, vec3 innormal, float bone_number, vec4 bones, out vec3 outposition, out vec3 outnormal) {
    vec4 position = vec4(inposition, 1);
    vec4 normal = vec4(innormal, 0);
    vec4 temp;

    mat4 bone0 = boneAtIndex(bones[0]);
    mat4 bone1 = boneAtIndex(bones[1]);
    mat4 bone2 = boneAtIndex(bones[2]);
    mat4 bone3 = boneAtIndex(bones[3]);

    temp = vec4(0);
    temp += bone0 * position;
    temp += bone1 * position;
    temp += bone2 * position;
    temp += bone3 * position;
    temp /= bone_number;
    outposition = vec3(temp);

    temp = vec4(0);
    temp += bone0 * normal;
    temp += bone1 * normal;
    temp += bone2 * normal;
    temp += bone3 * normal;
    outnormal = normalize(vec3(temp));
}

void main() {
    vec3 position, normal;

    transform(a_position, a_normal, a_bone_number, a_bones, position, normal);

    v_normal = normal;
    v_uv = a_uv + u_uv_offset.xy;

    gl_Position = u_mvp * vec4(position, 1);
}
`

SHADERS.wvsparticles = `
uniform mat4 u_mvp;
uniform vec2 u_dimensions;

attribute vec3 a_position;
attribute vec2 a_uva_rgb;

varying vec2 v_uv;
varying vec4 v_color;

void main() {
    vec3 uva = decodeFloat3(a_uva_rgb[0]);
    vec3 rgb = decodeFloat3(a_uva_rgb[1]);

    v_uv = uva.xy / u_dimensions;
    v_color = vec4(rgb, uva.z) / 255.0;

    gl_Position = u_mvp * vec4(a_position, 1);
}
`

SHADERS.wvsribbons = `
uniform mat4 u_mvp;
uniform vec3 u_uv_offset;

attribute vec3 a_position;
attribute vec2 a_uv;

varying vec2 v_uv;

void main() {
    v_uv = a_uv + u_uv_offset.xy;

    gl_Position = u_mvp * vec4(a_position, 1);
}
`

SHADERS.wvswhite = `
uniform mat4 u_mvp;

attribute vec3 a_position;
attribute vec4 a_bones;
attribute float a_bone_number;

void transform(vec3 inposition, float bone_number, vec4 bones, out vec3 outposition) {
    vec4 position = vec4(inposition, 1);
    vec4 temp;

    mat4 bone0 = boneAtIndex(bones[0]);
    mat4 bone1 = boneAtIndex(bones[1]);
    mat4 bone2 = boneAtIndex(bones[2]);
    mat4 bone3 = boneAtIndex(bones[3]);

    temp = vec4(0);
    temp += bone0 * position;
    temp += bone1 * position;
    temp += bone2 * position;
    temp += bone3 * position;
    temp /= bone_number;
    outposition = vec3(temp);
}

void main() {
    vec3 position;

    transform(a_position, a_bone_number, a_bones, position);

    gl_Position = u_mvp * vec4(position, 1);
}
```

**Comment by flowtsohg** [comment]
To be honest the Ruby compilation is mostly there because of the earlier design of the code base, where all of the code was one closure.
I still keep it because it let's me keep the handler files spread around with the before/after.js files, and because it allows to create the doc files properly (although I didn't regenerate them in quite some time, and there might be an easier way to do this either way).

Also you can't compress the shaders like that, which you might or might not care about.

Just out of interest, care to share what you're doing with this library? And was it easy or not to set it up? (No one really used it beside me)

**Comment by Meai** [comment]
What do you mean by compression? Gzip compression can still happen, and you could minify and combine all sources with google closure compiler, although I havent set that up here.
If you want I can try to make a pull request with my current setup. The files are a lot larger, especially the .mdx, but I dont even know how you currently work. I mean some of the files like after.js etc. are not valid .js right now, you would have to compile every time you change anything? 

Regarding the ease of setup: The main confusion came from this weird url object, I had no idea how or where my files need to be so that the library loads them correctly. Actually I still dont know exactly how it works, what's the "header" supposed to return? Apparently a .json file, but what format does it have to have and how are you generating that?

I also havent figured out what exactly the reason for BaseModel and BaseModelInstance is, I suppose you would use BaseModelInstance for models that dont have animations?
The reason I want to use this library is because I'm bad at 3d modelling/graphics and I want to start out with a few models so I can see how things work.
Btw, I saw you made a fix in the animation system, what does that do? Is it related to the issue about https://github.com/flowtsohg/mdx-m3-viewer/issues/8

Oh and I also dont know what this one means: https://github.com/flowtsohg/mdx-m3-viewer/issues/5
I looked at the model in the wc3viewer and it doesnt appear to consist of anything?
I guess I dont know a lot yet..

**Comment by flowtsohg** [comment]
Well, I am using my GLSL minifier to compress the shader source. I am not sure how noticeable it is after gzip, but whatever.
And yes, things like before.js and after.js are why it has to be compiled right now, however I am not sure if there's a better way to do this.

The urls argument should basically give paths at which files can be requested. This can be an actual path on your OS, or some PHP file that gets the correct file, or whatever - as long as the viewer requests for a specific MPQ file and gets it back.
For example, if you have all of the MPQs unpacked in a "WC3" folder that is in the same folder as your HTML page, the mpqFile could return `"WC3/" + file`.
The local file one is just if you want to load arbitrary local files that aren't related to WC3/SC2, such as the sky/grass/water textures, or anything else.
Now headers are a mechanism to allow loading custom resources. The ID is whatever you want, it completely depends on your site. It's really just an arbitrary identifier, that should end up in the server serving a JSON object that describes the resource to load.
E,g. if you have MyDragon.mdx in a Custom folder, you can decide that when the client asks for "MyDragon", it will serve a JSON object such as this:
```js
{"models": [{"url": "custom/MyDragon.mdx"}]}
```
Then, if you set up the header function to return `id` calling `viewer.load("MyDragon")` will end up calling `urls.header("MyDragon")`, which will return `"MyDragon"` which your server can use to get the above JSON file, and finally the library will load `"custom/MyDragon.mdx"`.

Base classes are just there because I had a lot of shared parts between the MDX and M3 handlers, so I mostly put them together.

The "fix" to the animation isn't related to #8 (which I'll probably never bother doing, because it really isn't required). It tries to fix some issues with keyframes, because nobody knows exactly how they work in-game. Specifically, how they work when a sequence doesn't have a beginning or ending keyframe.

The model referred to from #5 is there, it's just a UI model that is very tiny, so it can't be seen without a lot of zooming/scaling.

Particle emitters work fine last time I checked (which was a long time ago), but I do believe I messed something up with ubersplats.
It also might be possible that the keyframes don't work properly for that model, because of the above issue with keyframes.
